### DISCUSSION TITLE
Proposal: Use Java's Optional for implementing defaultValue and defaultExpression

### DISCUSSION THREAD

**Discussion by Chessray** [mapstruct/mapstruct#3364] [Ideas]
Currently, using `defaultValue` or `defaultExpression` results in generated Java code of the form:

`if (source != null) {
     mapToTarget(source);
}
else {
    mapToTarget(default);
}`

In functional spirit, this would look much nicer if generated as
`mapToTarget(Optional.ofNullable(source).orElseGet(() -> default));`

What does everyone think?

**Comment by filiphr** [comment]
Using optional for this will complicate things a bit.

The reason is that you might have custom presence checks methods and then our code looks like:

```java
if (isNotEmpty( source.getValue() ) {
    target.setValue( source.getValue() );
} else {
    target.setValue( <default> );
}
```

In addition to that I personally don't like using optional just to replace a null check. However, I am open to changing my mind if I see some examples where it looks OK.

@Chessray do you have some actual mapping examples that we can look into a bit better?
